service/storage_proxy.cc

```diff
@@ -5436,12 +5436,8 @@ storage_proxy::handle_read_digest(const rpc::client_info& cinfo, rpc::opt_time_p
 
 future<>
 storage_proxy::handle_truncate(rpc::opt_time_point timeout, sstring ksname, sstring cfname) {
-        return do_with(utils::make_joinpoint([] { return db_clock::now();}),
-                        [this, ksname, cfname](auto& tsf) {
-            return container().invoke_on_all(_write_smp_service_group, [ksname, cfname, &tsf](storage_proxy& sp) {
-                return sp._db.local().truncate(ksname, cfname, [&tsf] { return tsf.value(); });
-            });
-        });
+    slogger.info("Truncting {}.{}", ksname, cfname);
+    return replica::database::truncate_table_on_all_shards(_db, ksname, cfname, [ts = db_clock::now()] { return make_ready_future<db_clock::time_point>(ts); });
 }
```

Comment:
interesting, previously we could have different shards picking different timestamps, as the callback calculated a timestamp when called, not upfront.

example: shard s1 picks T1. shard s2 stalls, and picks a different timestamp T2, where T2 > T1.

I see the timestamp is used for creation of snapshot directory in database::truncate():

```
    db_clock::time_point truncated_at = co_await tsf();

    if (auto_snapshot) {
        auto name = format("{:d}-{}", truncated_at.time_since_epoch().count(), cf.schema()->cf_name());
        co_await cf.snapshot(*this, name);
    }
```

So previously, it could happen that snapshot could pick different directories in different shards, right?

You're making the code more robust here, and you're actually fixing a problem and an issue should be opened?!

Reply:
Turns out joinpoint provides the correct semantics. sorry for the noise.

Reply:
yes, but the caller could theoretically pass a timestamp_func that will return the current time on each shard, but that won't work, as you noticed.

joinpoint is a subtle and unvisible way of synchronizing between the shard via this callback.
I think it was a bad idea to use that here.